File: src/unity/Runtime/Ads/IAd.cs
using System;
using System.Threading.Tasks;

namespace EE {
    public class AdLoadResult {
        public string Network;
        public bool Result;
        public int ErrorCode;
        public string ErrorMessage;
    }

    public class AdObserver {
        /// <summary>
        /// Occurs when this ad is loaded.
        /// </summary>
        public Action OnLoaded { get; set; }

        /// <summary>
        /// Occurs when this ad finishes loading.
        /// </summary>
        public Action<AdLoadResult> OnLoadResult { get; set; }

        /// <summary>
        /// Occurs when user clicks this ad.
        /// </summary>
        public Action OnClicked { get; set; }
    }

    public interface IAd : IObserverManager<AdObserver> {
        /// <summary>
        /// Destroys this ad.
        /// </summary>
        void Destroy();

        /// <summary>
        /// Checks whether this ad is loaded.
        /// </summary>
        bool IsLoaded { get; }

        /// <summary>
        /// Asynchronously loads this ad.
        /// </summary>
        Task<bool> Load();
    }
}

File: src/ts/src/ads/IAd.ts
import { IObserverManager } from "../core";

export interface AdLoadResult {
    network: string,
    result: boolean,
    errorCode: number,
    errorMessage: string,
}

export interface AdObserver {
    onLoaded?(): void;
    onLoadResult?(result: AdLoadResult): void
    onClicked?(): void;
}

export interface IAd extends IObserverManager<AdObserver> {
    destroy(): void;
    isLoaded: boolean;
    load(): Promise<boolean>;
}